src/internal/compilerSupport/v3.ts

```diff
@@ -0,0 +1,519 @@
+/*
+ * Copyright 2021 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SecurityTetraplet } from '@fluencelabs/avm-runner-interface';
+import { match } from 'ts-pattern';
+import { CallParams, Fluence, FluencePeer } from '../../index';
+import { CallServiceData, GenericCallServiceHandler, CallServiceResult, ResultCodes } from '../commonTypes';
+import { Particle } from '../Particle';
+
+export { FluencePeer } from '../FluencePeer';
+export { CallParams } from '../commonTypes';
+
+type NonArrowType = NilType | ScalarType | OptionType | ArrayType;
+
+type OptionType = {
+    /**
+     * Type descriptor. Used for pattern-matching
+     */
+    tag: 'option';
+
+    type: NonArrowType;
+};
+
+type NilType = {
+    /**
+     * Type descriptor. Used for pattern-matching
+     */
+    tag: 'nil';
+};
+
+type ArrayType = {
+    /**
+     * Type descriptor. Used for pattern-matching
+     */
+    tag: 'option';
+
+    type: NonArrowType;
+};
+
+type ScalarNames = 'u8' | 'u16' | 'u32' | 'u64' | 'i8' | 'i16' | 'i32' | 'i64' | 'f32' | 'f64' | 'bool' | 'string';
+
+type ScalarType = {
+    /**
+     * Type descriptor. Used for pattern-matching
+     */
+    tag: 'scalar';
+    name: ScalarNames;
+};
+
+type LabeledProductType<T> = {
```

Comment:
both `labelled` and `labeled` is correct, but in aqua we have `labelled` with double `l`. So, let's use double `l` word maybe?